### ISSUE TITLE
Only first forwarded port is forwarded when container is on an ipv6 network

### ISSUE THREAD

**Issue by ansel1** [docker/for-mac#5247]
- [x] I have tried with the latest version of Docker Desktop
  - [x] I have tried disabling enabled experimental features
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: EF4CD0CE-9C35-403C-9E2C-28807F9AA87A/20210113194324

### Expected behavior

Create an ipv6 network, and launch a container on that network with two ports forwarded.  Only one port (randomly chosen) will get forwarded.

    docker network create --ipv6 --gateway fd4a:1b2e:4486:8e48::1 --subnet fd4a:1b2e:4486:8e48::/64 testv6
    docker run --rm -p 5453:5432 -p 443:8443 --network testv6 -d nginx
    netstat -an | grep LISTEN | grep 443
    netstat -an | grep LISTEN | grep 5453

I expect netstat to show that both ports are listening.  `nc -vz localhost <port>` can also be used to test the forwarded port.

### Actual behavior

Only one port will be listening.  It's random which one it is.  Here's sample output:

```
❯ docker ps
CONTAINER ID   IMAGE     COMMAND                  CREATED       STATUS       PORTS                                                                                        NAMES
a1fae14a80b5   nginx     "/docker-entrypoint.…"   2 hours ago   Up 2 hours   80/tcp, 0.0.0.0:5453->5432/tcp, :::5453->5432/tcp, 0.0.0.0:443->8443/tcp, :::443->8443/tcp   charming_kalam
❯ nc -vz localhost 5453
Connection to localhost port 5453 [tcp/surebox] succeeded!
❯ nc -vz localhost 443
nc: connectx to localhost port 443 (tcp) failed: Connection refused
nc: connectx to localhost port 443 (tcp) failed: Connection refused
```

### Information

- `docker ps` indicates it is forwarding both ports.
- This is reproducible 100% on all macs we're tried.
- It only happens if the network is ipv6
- It does not happen on a native linux host.  Haven't tried it with docker for windows.
- We think the problem is recent.  We first noticed some ports weren't being forward around Jan 8. 
- macOS Version: Big Sur 11.1

**Comment by colathro** [comment]
I believe I am hitting this also. 

I'm following a guide on setting up Service Fabric outlined here: https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-get-started-mac

It seems very inconsistent on which port is chosen when the container starts.

"docker run --name sftestcluster -d -v /var/run/docker.sock:/var/run/docker.sock -p 19080:19080 -p 19000:19000 -p 25100-25200:25100-25200 mcr.microsoft.com/service-fabric/onebox:u18"

While I don't setup a new network, the setup guide has you adjust docker daemon for IPv6: 
{
    "ipv6": true,
    "fixed-cidr-v6": "fd00::/64"
}

**Comment by colathro** [comment]
I downgraded docker desktop to 3.0.3 and the container with multiple published ports works as intended.

**Comment by bartsmykla** [comment]
Same problem here. Downgrading to 3.0.3 solved it.

**Comment by ansel1** [comment]
Still present in 3.1.0

**Comment by stephen-turner** [comment]
Interesting. We know ipv6 isn't fully working, but we don't intend to regress anything that was previously working. We took a new engine from upstream between 3.0.3 and 3.0.4, so I wonder if that is to blame.

**Comment by djs55** [comment]
To narrow down the problem for us, could you try accessing the exposed ports from inside a container with `--net=host`.

For example if I expose port 80:
```
$ docker run -p 80:80 -d nginx
```
and then
```
$ docker run -it --net=host alpine sh
/ # wget http://127.0.0.1:80
Connecting to 127.0.0.1:80 (127.0.0.1:80)
saving to 'index.html'
index.html           100% |**************************************************************************************************|   612  0:00:00 ETA
'index.html' saved
```
If the port works inside the container, it suggests a problem with the exposure to the host. If it doesn't work inside the container, it suggests a problem with the engine.

**Comment by colathro** [comment]
@djs55 
Here is the output from the above commands:
```
Last login: Sun Jan 17 11:31:33 on ttys002
➜  ~ docker run -p 80:80 -d nginx
Unable to find image 'nginx:latest' locally
latest: Pulling from library/nginx
a076a628af6f: Pull complete 
0732ab25fa22: Pull complete 
d7f36f6fe38f: Pull complete 
f72584a26f32: Pull complete 
7125e4df9063: Pull complete 
Digest: sha256:10b8cc432d56da8b61b070f4c7d2543a9ed17c2b23010b43af434fd40e2ca4aa
Status: Downloaded newer image for nginx:latest
93953a3c386aef98c589738b71fb6b7bd514fbcb3e91df5a4f57af02bee7e3f7
➜  ~ docker ps
CONTAINER ID   IMAGE     COMMAND                  CREATED          STATUS          PORTS                NAMES
93953a3c386a   nginx     "/docker-entrypoint.…"   20 seconds ago   Up 18 seconds   0.0.0.0:80->80/tcp   pensive_goldwasser
➜  ~ docker run -it --net=host alpine sh
Unable to find image 'alpine:latest' locally
latest: Pulling from library/alpine
596ba82af5aa: Pull complete 
Digest: sha256:d9a7354e3845ea8466bb00b22224d9116b183e594527fb5b6c3d30bc01a20378
Status: Downloaded newer image for alpine:latest
/ # wget http://127.0.0.1:80
Connecting to 127.0.0.1:80 (127.0.0.1:80)
saving to 'index.html'
index.html           100% |********************************|   612  0:00:00 ETA
'index.html' saved
/ # 
```

**Comment by ansel1** [comment]
It works inside the container, but not from host:

```
❯ docker network create --ipv6 --gateway fd4a:1b2e:4486:8e48::1 --subnet fd4a:1b2e:4486:8e48::/64 testv6
ce22bee018df60f5e93128885e7ef205e1289f0fdabed59724b075cb936f3c1c

❯ docker run --rm -p 80:80 -p 81:80 --network testv6 -d nginx
1b959cf06236561994a09cb8d214beb409a478920774d91b1159ef5717882f02

❯ nc -vz 127.0.0.1 80
nc: connectx to 127.0.0.1 port 80 (tcp) failed: Connection refused

❯ nc -vz 127.0.0.1 81
Connection to 127.0.0.1 port 81 [tcp/hosts2-ns] succeeded!

❯ docker run -it --net=host alpine sh
/ # nc -vz 127.0.0.1 80
127.0.0.1 (127.0.0.1:80) open
/ # nc -vz 127.0.0.1 81
127.0.0.1 (127.0.0.1:81) open
```

I also tried using "localhost" instead of "127.0.0.1", same result.

**Comment by sethp-verica** [comment]
I'm seeing this, also. When I run `docker run --rm -p 1234:1234 -p 4321:4321 --net=testv6 k8s.gcr.io/pause:3.2` I see the following in `~/Library/Containers/com.docker.docker/Data/log/host/com.docker.backend.log` that may be relevant:

```
time="2021-01-19T14:53:28-08:00" level=info msg="grpcfuseClient.Add(39b592085f64f898fa117c029d37ee4351e1282946b950a18ef1bb928cfc2fe5, /intelligent_chandrasekhar, [])"
time="2021-01-19T14:53:28-08:00" level=info msg="vpnkitExposer.Add(39b592085f64f898fa117c029d37ee4351e1282946b950a18ef1bb928cfc2fe5, /intelligent_chandrasekhar, [TCP 0.0.0.0:1234 -> 127.0.0.1:1234 T
CP :::1234 -> :::1234 TCP 0.0.0.0:4321 -> 127.0.0.1:4321 TCP :::4321 -> :::4321])"
time="2021-01-19T14:53:28-08:00" level=info msg="adding docker-containers/39b592085f64f898fa117c029d37ee4351e1282946b950a18ef1bb928cfc2fe5 tcp forward from 0.0.0.0:1234 to 127.0.0.1:1234"
time="2021-01-19T14:53:28-08:00" level=info msg="adding docker-containers/39b592085f64f898fa117c029d37ee4351e1282946b950a18ef1bb928cfc2fe5 tcp forward from :::1234 to :::1234"
time="2021-01-19T14:53:28-08:00" level=error msg="unable to expose port TCP :::1234 -> :::1234: listen tcp [::]:1234: bind: address already in use"
goroutine 14 [running]:
github.com/docker/pinata/common/pkg/ports.(*vpnkitExposer).Add(0xc0003f9da0, 0x1ae23cab, 0xf37a0bcd0, 0x5f9c5a0, 0xc00083a180, 0x40, 0xc0002ef880, 0x1a, 0xc00054d5e0, 0x4, ...)
        /Users/administrator/jenkins/workspace/desktop_desktop-build_master/src/github.com/docker/pinata/common/pkg/ports/vpnkit.go:55 +0x5b8
github.com/docker/pinata/common/pkg/docker.(*portResource).Start(0xc000715cc0, 0x1ae23cab, 0xf37a0bcd0, 0x5f9c5a0, 0xc00083a180, 0x40)
        /Users/administrator/jenkins/workspace/desktop_desktop-build_master/src/github.com/docker/pinata/common/pkg/docker/port.go:265 +0x1ec
github.com/docker/pinata/common/pkg/docker.(*Resyncer).watch(0xc00010a780, 0x55a7280, 0xc0000d4008, 0xc000651c80, 0xc0007b6c60, 0x0, 0x0)
        /Users/administrator/jenkins/workspace/desktop_desktop-build_master/src/github.com/docker/pinata/common/pkg/docker/resyncer.go:165 +0xba6
github.com/docker/pinata/common/pkg/docker.(*Resyncer).syncUntilCancelOrTimeout(0xc00010a780, 0x0, 0x0)
        /Users/administrator/jenkins/workspace/desktop_desktop-build_master/src/github.com/docker/pinata/common/pkg/docker/resyncer.go:236 +0x166
github.com/docker/pinata/common/pkg/docker.(*Resyncer).Start(0xc00010a780)
        /Users/administrator/jenkins/workspace/desktop_desktop-build_master/src/github.com/docker/pinata/common/pkg/docker/resyncer.go:202 +0x70
created by main.startPortForwarding
        /Users/administrator/jenkins/workspace/desktop_desktop-build_master/src/github.com/docker/pinata/common/cmd/com.docker.backend/main.go:166 +0x296
time="2021-01-19T14:53:28-08:00" level=error msg="opening a port forward for 39b592085f64f898fa117c029d37ee4351e1282946b950a18ef1bb928cfc2fe5: listen tcp [::]:1234: bind: address already in use"
goroutine 14 [running]:
github.com/docker/pinata/common/pkg/docker.(*portResource).Start(0xc000715cc0, 0x1ae23cab, 0xf37a0bcd0, 0x5f9c5a0, 0xc00083a180, 0x40)
        /Users/administrator/jenkins/workspace/desktop_desktop-build_master/src/github.com/docker/pinata/common/pkg/docker/port.go:266 +0x2e5
github.com/docker/pinata/common/pkg/docker.(*Resyncer).watch(0xc00010a780, 0x55a7280, 0xc0000d4008, 0xc000651c80, 0xc0007b6c60, 0x0, 0x0)
        /Users/administrator/jenkins/workspace/desktop_desktop-build_master/src/github.com/docker/pinata/common/pkg/docker/resyncer.go:165 +0xba6
github.com/docker/pinata/common/pkg/docker.(*Resyncer).syncUntilCancelOrTimeout(0xc00010a780, 0x0, 0x0)
        /Users/administrator/jenkins/workspace/desktop_desktop-build_master/src/github.com/docker/pinata/common/pkg/docker/resyncer.go:236 +0x166
github.com/docker/pinata/common/pkg/docker.(*Resyncer).Start(0xc00010a780)
        /Users/administrator/jenkins/workspace/desktop_desktop-build_master/src/github.com/docker/pinata/common/pkg/docker/resyncer.go:202 +0x70
created by main.startPortForwarding
        /Users/administrator/jenkins/workspace/desktop_desktop-build_master/src/github.com/docker/pinata/common/cmd/com.docker.backend/main.go:166 +0x296
```

**Comment by StefanScherer** [comment]
Thanks @ansel1 for the repro steps, we're investigating.

**Comment by StefanScherer** [comment]
We have found the issue and could fix it in our codebase. It will be available in the next release of Docker Desktop.

**Comment by stephen-turner** [comment]
Reopening until it's released.

**Comment by stephen-turner** [comment]
This is now fixed in versions 3.2.0 and 3.2.1. Closing the ticket now.
